test/e2e/azure_vmssextensions.go

```diff
@@ -0,0 +1,111 @@
+//go:build e2e
+// +build e2e
+
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2021-11-01/compute"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	infrav1exp "sigs.k8s.io/cluster-api-provider-azure/exp/api/v1beta1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// AzureVMSSExtensionsSpecInput is the input for AzureVMSSExtensionsSpec.
+type AzureVMSSExtensionsSpecInput struct {
+	BootstrapClusterProxy framework.ClusterProxy
+	Namespace             *corev1.Namespace
+	VMName                string
+	ClusterName           string
+}
+
+// AzureVMSSExtensionsSpec implements a test that verifies VM extensions are created and deleted.
+func AzureVMSSExtensionsSpec(ctx context.Context, inputGetter func() AzureVMSSExtensionsSpecInput) {
```

Comment:
Since this has some overlapping code with the VM version of this, it might make sense to combine them into one spec. Since we're ranging over machines and machinepools it will simply skip those if the cluster has none.

Reply:
Ah I didn't think about that! There might still be a lot of duplicated code since machines and machinepools use different types, but it does make sense to combine them into one spec.